torch/_inductor/subgraph_lowering.py

```diff
@@ -0,0 +1,142 @@
+"""Utilities for lowering subgraphs used by higher order operators
+
+"""
+
+import functools
+import operator
+from dataclasses import dataclass
+from typing import List, TypeVar
+
+import torch
+
+from . import ir
+from .exc import SubgraphLoweringException
+from .ops_handler import SimpleCSEHandler
+from .virtualized import ops, V, WrapperHandler
+
+T = TypeVar("T")
+
+
+class PointwiseSubgraphLowering(torch.fx.Interpreter):
+    graph_outputs: List[ir.IRNode]
+
+    def __init__(
+        self,
+        gm: torch.fx.GraphModule,
+        root_graph_lowering: "torch._inductor.graph.GraphLowering",
+    ):
+        super().__init__(gm)
+        self.graph_outputs = []
+        self.root_graph = root_graph_lowering
+
+    @property
+    def sizevars(self):
+        return self.root_graph.sizevars
+
+    def mark_buffer_mutated(self, name):
+        raise SubgraphLoweringException("Mutations are not supported in this context")
+
+    def register_buffer(self, data):
+        raise SubgraphLoweringException(
+            "Buffer creation is not supported in this context"
+        )
+
+    def call_function(self, target, args, kwargs):
+        from .lowering import lowerings
+
+        if target is operator.getitem and isinstance(args[0], (list, tuple, dict)):
+            return super().call_function(target, args, kwargs)
+
+        assert isinstance(target, torch._ops.OpOverload)
+
+        if target not in lowerings:
+            raise SubgraphLoweringException(
+                f"{target} not supported in subgraph, (missing lowering)"
+            )
+
+        if torch.Tag.pointwise not in target.tags:
+            raise SubgraphLoweringException(
+                f"Only pointwise operators are supported in this context, but got {target}"
+            )
+
+        return lowerings[target](*args, **kwargs)
+
+    def output(self, target, args, kwargs):
+        self.graph_outputs.extend(args)
+
+
+@dataclass
+class InputDescriptor:
+    dtype: torch.dtype
+    device: torch.device
+
+
+class TracingOpsHandler(WrapperHandler[T]):
+    def __init__(self, tracer, num_inputs):
+        parent = tracer.create_proxy("placeholder", "ops", (), {})
+        super().__init__(parent)
+        self.tracer = tracer
+
+        self.placeholders = [
+            self.tracer.create_proxy("placeholder", f"input{i}", (), {})
+            for i in range(num_inputs)
+        ]
+
+    def placeholder(self, idx):
+        return self.placeholders[idx]
+
+    def output(self, *args):
+        return self.tracer.create_node(
+            "output", "output", tuple(self.tracer.create_arg(a) for a in args), {}
+        )
+
+
+def lower_pointwise_subgraph(subgraph: ir.Subgraph, inputs: List[InputDescriptor]):
```

Comment:
Not against it, but why is this needed? For templated attention we just iterated and manually called lowerings - is the main reason to check various kinds of safety issues?

Reply:
Yes it's mainly for safety, I don't want the lowerings being called to interact with the top level graph in any way. So I isolate it by overriding `V.graph`.

For templated attention if it's not a pure pointwise operation then this function won't work, but the  `PointwiseSubgraphLowering` class could easily be adapted into `SubgraphLowering` that isn't as strict about what functions are allowed.

Reply:
I mean, it's an indirect load, which is kind of a pointwise function?

Reply:
Yeah there's no conceptual problem, it's just that I can't lower an `ops.load` inside a helper function in triton so want to ban such cases in this functions.